File: Leetcode/94-Binary-Tree-Inorder-Traversal/BinaryTreeInorderTraversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderTraversal(root *TreeNode) []int {
	res := make([]int, 0)
	cur := root
	var prev *TreeNode
	for cur != nil {
		if cur.Left == nil {
			res = append(res, cur.Val)
			cur = cur.Right
			continue
		}
		prev = cur.Left
		for prev.Right != nil && prev.Right != cur {
			prev = prev.Right
		}
		if prev.Right == nil {
			prev.Right, cur = cur, cur.Left
		} else {
			res = append(res, cur.Val)
			prev.Right, cur = nil, cur.Right
		}
	}
	return res
}

File: Leetcode/94-Binary-Tree-Inorder-Traversal/BinaryTreeInorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        TreeNode* cur = root;
        TreeNode* prev = NULL;
        while (cur!=NULL) {
            if (cur->left == NULL) {
                res.push_back(cur->val);
                cur = cur->right;
                continue;
            }
            prev = cur->left;
            while (prev->right!=NULL && prev->right!=cur) {
                prev = prev->right;
            }
            if (prev->right == NULL) {
                prev->right = cur;
                cur = cur->left;
            } else {
                prev->right = NULL;
                res.push_back(cur->val);
                cur = cur->right;
            }
        }
        return res;
    }
};
